src/chainbin/nnet3-chain-get-egs.cc

```diff
@@ -30,6 +30,60 @@
 namespace kaldi {
 namespace nnet3 {
 
+/** add iVectors to Egs.
+*/
+static void GetIvectorsForEgs(int32 t, int32 left_context,
+                              int32 right_context, int32 frames_per_eg,
+                              int32 ivector_period,
+                              const MatrixBase<BaseFloat> *ivector_feats,
+                              NnetChainExample *eg) {
+  KALDI_ASSERT(ivector_feats->NumRows() > 0);
+  if (ivector_period == 0) {
+    // single ivector case.
+    // if applicable, add the iVector feature.
+    // try to get closest frame to middle of window to get
+    // a representative iVector.
+    int32 closest_frame = t + frames_per_eg / 2;
+    KALDI_ASSERT(ivector_feats->NumRows() > 0);
+    if (closest_frame >= ivector_feats->NumRows())
+      closest_frame = ivector_feats->NumRows() - 1;
+    Matrix<BaseFloat> ivector(1, ivector_feats->NumCols());
+    ivector.Row(0).CopyFromVec(ivector_feats->Row(closest_frame));
+    NnetIo ivector_io("ivector", 0, ivector);
+    eg->inputs[1].Swap(&ivector_io);
+  } else {
+    // multiple ivectors case.
+    // try to get a representative iVector every ivector_period frames
+    // in an eg.
+    // num_ivectors_in_left_context is the num of ivectors for frames whose
+    // t < 0. It is used to compute the value of closest_frame.
+    // num_ivectors is the num of ivectors for frames whose t >= 0.
+    // Both of them are computed according to how Round descriptor works, which
+    // basically returns floor(t / <t-modulus>) * <t-modulus>.
+    int32 num_ivectors_in_left_context =
+        -1 * DivideRoundingDown(-left_context, ivector_period); 
+    int32 num_ivectors = num_ivectors_in_left_context + 1 +
+        DivideRoundingDown(frames_per_eg + right_context - 1, ivector_period);
+    Matrix<BaseFloat> ivectors(num_ivectors, ivector_feats->NumCols());
+    for (int32 n = 0; n < num_ivectors; n++) {
+      // closest_frame is the frame at whole multiples of ivector_period
+      int32 closest_frame = t -
```

Comment:
Yes, it looks like you're right.  OK.


On Sat, Mar 26, 2016 at 7:15 PM, Yiming Wang <notifications@github.com>
wrote:

> In src/chainbin/nnet3-chain-get-egs.cc
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r57520766>:
>
> > +    // multiple ivectors case.
> > +    // try to get a representative iVector every ivector_period frames
> > +    // in an eg.
> > +    // num_ivectors_in_left_context is the num of ivectors for frames whose
> > +    // t < 0. It is used to compute the value of closest_frame.
> > +    // num_ivectors is the num of ivectors for frames whose t >= 0.
> > +    // Both of them are computed according to how Round descriptor works, which
> > +    // basically returns floor(t / <t-modulus>) * <t-modulus>.
> > +    int32 num_ivectors_in_left_context =
> > +        -1 * DivideRoundingDown(-left_context, ivector_period);
> > +    int32 num_ivectors = num_ivectors_in_left_context + 1 +
> > +        DivideRoundingDown(frames_per_eg + right_context - 1, ivector_period);
> > +    Matrix<BaseFloat> ivectors(num_ivectors, ivector_feats->NumCols());
> > +    for (int32 n = 0; n < num_ivectors; n++) {
> > +      // closest_frame is the frame at whole multiples of ivector_period
> > +      int32 closest_frame = t -
>
> Regarding ivector_feats:
>
> in steps/nnet3/get_egs.sh, it seems that ivector_feats is the results of
> repeating each feature in ivector_online.scp ivector_period of times (with
> negative value for --n option for subsample-feats in the string variable
> "ivector_opt")
> So I think after that ivec_feats has the same length as the utterance. So
> we don't need to divide ivector_frame by ivector_period?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files/dba966becb40d75369b45732f575e1f900987824#r57520766>
>